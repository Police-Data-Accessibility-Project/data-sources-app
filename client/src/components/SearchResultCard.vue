<template>
	<GridItem
		class="flex flex-col border border-neutral-400 p-3 text-lg leading-snug"
		data-test="search-result-card"
	>
		<h2 class="text-xl font-semibold line-clamp-2" data-test="search-result-title">
			{{ dataSource.data_source_name }}
		</h2>
		<p
			class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"
			data-test="search-result-record-label"
		>
			Record type
		</p>
		<div
			data-test="search-result-record-type"
			v-if="dataSource.record_type"
			class="mt-1 py-[.125rem] px-3 rounded-full bg-brand-wine/10 dark:bg-brand-wine dark:text-white w-fit"
		>
			{{ dataSource.record_type }}
		</div>
		<p 
			v-else 
			data-test="search-result-record-type-unknown"
		>
			Unknown
		</p>
		<div class="search-result-agency" data-test="search-result-agency">
			<p
			class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"
			data-test="search-result-record-label"
			>
			Agency
		</p>
			<p v-if="dataSource.agency_name" data-test="search-result-agency-known">
				{{ dataSource.agency_name }}
			</p>
			<p v-else data-test="search-result-agency-unknown">Unknown</p>
		</div>
<!-- hiding place for now
		<div data-test="search-result-place">
			<p
				v-if="dataSource.municipality && dataSource.state_iso"
				data-test="search-result-place-state-municipality"
			>
				{{ dataSource.municipality }}, {{ dataSource.state_iso }}
			</p>
			<p v-else-if="dataSource.municipality" data-test="search-result-place-municipality">
				{{ dataSource.municipality }}
			</p>
			<p v-else-if="dataSource.state_iso" data-test="search-result-place-state">
				{{ dataSource.state_iso }}
			</p>
			<p v-else data-test="search-result-place-unknown">Location Unknown</p>
		</div>
-->
		<p
			class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"
			data-test="search-result-label-coverage"
		>
			Time range
		</p>
		<p
			v-if="dataSource.coverage_start && dataSource.coverage_end"
			data-test="search-result-coverage-start-end"
		>
			{{ formatDate(dataSource.coverage_start) }}&ndash;{{ formatDate(dataSource.coverage_end) }}
		</p>
		<p
			v-else-if="dataSource.coverage_start && !dataSource.coverage_end"
			data-test="search-result-coverage-start"
		>
			{{ formatDate(dataSource.coverage_start) }}&ndash;Unknown end
		</p>
		<p
			v-else-if="!dataSource.coverage_start && dataSource.coverage_end"
			data-test="search-result-coverage-end"
		>
			Unknown start&ndash;{{ formatDate(dataSource.coverage_end) }}
		</p>
		<p v-else data-test="search-result-coverage-unknown">
			Unknown
		</p>
		<p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" data-test="search-result-label-formats">Formats available</p>
		<ul
			v-if="dataSource.record_format" 
			data-test="search-result-formats"
			class="mb-4"
		>
			<li
				:key="recordFormat"
				v-for="recordFormat in dataSource.record_format"
				data-test="search-result-format"
				class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit"
			>
				{{ recordFormat }}
			</li>
		</ul>
		<p data-test="search-result-format-unknown" v-else>
			Unknown
		</p>
		<div class="mt-auto flex flex-col">
			<Button
				:class="['text-lg font-medium px-4 py-1 mt-4 mb-2 lg:mx-0 max-w-full']"
				:disabled="!dataSource.source_url || dataSource.url_status === 'broken' || dataSource.url_status === 'not found'"
				@click="dataSource.source_url && dataSource.url_status !== 'broken' && dataSource.url_status !== 'not found' ? openSource() : null"
				data-test="search-result-source-button"
			>
				Visit data source <i class="fa fa-external-link"></i>
			</Button>
			<Button
				@click="showDetails"
				intent="secondary"
				class="text-lg font-medium px-4 py-1 lg:mx-0 max-w-full"
				data-test="search-result-source-details-button"
			>
				View details
			</Button>
		</div>
	</GridItem>
</template>

<script>
import { Button, GridItem } from 'pdap-design-system';
import formatDateForSearchResults from '../util/formatDate'

export default {
	name: 'SearchResultCard',
	components: {
		Button,
		GridItem,
	},
	props: {
		dataSource: Object,
	},
	methods: {
		showDetails() {
			this.$router.push(`/data-sources/${this.dataSource.airtable_uid}`);
		},
		openSource() {
			window.open(this.dataSource.source_url, '_blank');
		},
		formatDate: formatDateForSearchResults
	},
};
</script>