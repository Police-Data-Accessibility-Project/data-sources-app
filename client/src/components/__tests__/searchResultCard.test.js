import SearchResultCard from '../SearchResultCard.vue';
import { mount } from '@vue/test-utils';
import { describe, it, expect, beforeEach, vi } from 'vitest';
import { nextTick } from 'vue';

let wrapper;

// const $routerMock = vi.mock("vue-router");
const push = vi.fn();
const $routerMock = {
	push,
};

describe('SearchResultCard with all data', () => {
	const dataSource = {
		agency_name: 'Cicero Police Department - IN',
		agency_supplied: false,
		airtable_uid: 'rec7edsmcuVQwMMXG',
		coverage_end: '2018-01-01',
		coverage_start: '2017-01-01',
		data_source_name: 'Calls for Service for Cicero Police Department - IN',
		municipality: 'Bridgeview',
		name: 'Cicero Police Department - IN',
		record_format: "['json', 'pdf']",
		record_type: 'Calls for Service',
		source_url:
			'https://cityprotect.com/agency/cc0d9a3d-50b2-4424-ae25-5699ba6eaff9',
		state_iso: 'IN',
	};

	beforeEach(() => {
		wrapper = mount(SearchResultCard, {
			props: { dataSource },
			global: {
				mocks: {
					$router: $routerMock,
				},
			},
		});

		vi.unstubAllGlobals();
	});

	it('search result card exists with full data', () => {
		expect(wrapper.find('[data-test="search-result-card"]').exists()).toBe(
			true,
		);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('search result card contains a title', () => {
		expect(wrapper.get('[data-test="search-result-title"]').text()).toBe(
			dataSource.data_source_name,
		);
	});

	it('search result card contains an agency section', () => {
		expect(wrapper.find('[data-test="search-result-agency"]').exists()).toBe(
			true,
		);
	});

	it('search result card contains correct agency name', () => {
		expect(wrapper.get('[data-test="search-result-agency-known"]').text()).toBe(
			dataSource.agency_name,
		);
	});
	/* hiding place for now
	it('search result card contains a place section', () => {
		expect(wrapper.find('[data-test="search-result-place"]').exists()).toBe(
			true
		);
	});

	it('search result card contains correct place format', () => {
		expect(
			wrapper.get('[data-test="search-result-place-state-municipality"]').text()
		).toBe(`${dataSource.municipality}, ${dataSource.state_iso}`);
	});
*/
	it('search result card contains a record section', () => {
		expect(
			wrapper.find('[data-test="search-result-record-label"]').exists(),
		).toBe(true);
	});

	it('search result card contains correct record type', () => {
		expect(wrapper.get('[data-test="search-result-record-type"]').text()).toBe(
			dataSource.record_type,
		);
	});

	it('search result card contains a coverage date section', () => {
		expect(
			wrapper.find('[data-test="search-result-label-coverage"]').exists(),
		).toBe(true);
	});

	it('search result card contains correct coverage start and end date', () => {
		expect(
			wrapper.get('[data-test="search-result-coverage-start-end"]').text(),
		).toBe(
			`${wrapper.vm.formatDate(
				dataSource.coverage_start,
			)}â€“${wrapper.vm.formatDate(dataSource.coverage_end)}`,
		);
	});

	it('search result card contains a record formats section', () => {
		expect(
			wrapper.find('[data-test="search-result-label-formats"]').exists(),
		).toBe(true);
	});

	it('search result card contains a record formats div for each format', () => {
		expect(wrapper.find('[data-test="search-result-formats"]').exists()).toBe(
			true,
		);
	});

	it('Visit source button exists and calls window navigation with correct value on click', async () => {
		const spy = vi.spyOn(window, 'open');

		const button = wrapper.find(
			'[data-test="search-result-visit-source-button"]',
		);

		expect(button.exists()).toBe(true);

		await button.trigger('click');

		await nextTick();

		expect(spy).toHaveBeenCalledOnce();
		expect(spy).toHaveBeenCalledWith(dataSource.source_url, '_blank');
	});

	it('View details button exists and calls router.push with correct value on click', async () => {
		const button = wrapper.find(
			'[data-test="search-result-source-details-button"]',
		);

		expect(button.exists()).toBe(true);

		await button.trigger('click');

		await nextTick();

		expect(push).toHaveBeenCalledOnce();
		expect(push).toHaveBeenCalledWith(
			`/data-sources/${dataSource.airtable_uid}`,
		);
	});
});

describe('SearchResultCard with missing data', () => {
	const dataSource = {
		agency_name: null,
		agency_supplied: false,
		airtable_uid: 'rec7edsmcuVQwMMXG',
		coverage_end: null,
		coverage_start: null,
		data_source_name: 'Calls for Service for Cicero Police Department - IN',
		municipality: null,
		name: null,
		record_format: null,
		record_type: null,
		source_url: null,
		state_iso: null,
	};

	beforeEach(() => {
		wrapper = mount(SearchResultCard, {
			props: { dataSource },
		});
	});

	it('search result card exists with missing data', () => {
		expect(wrapper.find('[data-test="search-result-card"]').exists()).toBe(
			true,
		);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('search result card contains unknown for agency name', () => {
		expect(
			wrapper.find('[data-test="search-result-agency-unknown"]').exists(),
		).toBe(true);
	});

	// it("search result card contains unknown for place", () => {
	// 	expect(
	// 		wrapper.find('[data-test="search-result-place-unknown"]').exists(),
	// 	).toBe(true);
	// });

	it('search result card contains unknown for record type', () => {
		expect(
			wrapper.find('[data-test="search-result-record-type-unknown"]').exists(),
		).toBe(true);
	});

	it('search result card contains unknown for coverage', () => {
		expect(
			wrapper.find('[data-test="search-result-coverage-unknown"]').exists(),
		).toBe(true);
	});

	it('search result card contains unknown for record formats', () => {
		expect(
			wrapper.find('[data-test="search-result-format-unknown"]').exists(),
		).toBe(true);
	});
});

// describe("SearchResultCard with municipality but not state", () => {
// 	const dataSource = {
// 		municipality: "Bridgeview",
// 		state_iso: null,
// 	};

// 	wrapper = mount(SearchResultCard, {
// 		props: { dataSource },
// 	});

// 	it("search result card has only municipality", () => {
// 		expect(
// 			wrapper.get('[data-test="search-result-place-municipality"]').text(),
// 		).toBe(dataSource.municipality);
// 	});
// });

// describe("SearchResultCard with state but not municipality", () => {
// 	const dataSource = {
// 		municipality: null,
// 		state_iso: "IN",
// 	};

// 	wrapper = mount(SearchResultCard, {
// 		props: { dataSource },
// 	});

// 	it("search result card has only state", () => {
// 		expect(wrapper.get('[data-test="search-result-place-state"]').text()).toBe(
// 			dataSource.state_iso,
// 		);
// 	});
// });

describe('SearchResultCard with coverage start but not end', () => {
	const dataSource = {
		coverage_start: '2017-01-01',
		coverage_end: null,
	};

	beforeEach(() => {
		wrapper = mount(SearchResultCard, {
			props: { dataSource },
		});
	});

	it('search result card exists with coverage start but not end', () => {
		expect(wrapper.find('[data-test="search-result-card"]').exists()).toBe(
			true,
		);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('search result card has only coverage start date', () => {
		expect(
			wrapper.get('[data-test="search-result-coverage-start"]').text(),
		).toContain(wrapper.vm.formatDate(dataSource.coverage_start));
	});
});

describe('SearchResultCard with coverage end but not start', () => {
	const dataSource = {
		coverage_start: null,
		coverage_end: '2017-01-01',
	};

	beforeEach(() => {
		wrapper = mount(SearchResultCard, {
			props: { dataSource },
		});
	});

	it('search result card exists with coverage end but not start', () => {
		expect(wrapper.find('[data-test="search-result-card"]').exists()).toBe(
			true,
		);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('search result card has only coverage end date', () => {
		expect(
			wrapper.get('[data-test="search-result-coverage-end"]').text(),
		).toContain(wrapper.vm.formatDate(dataSource.coverage_end));
	});
});
