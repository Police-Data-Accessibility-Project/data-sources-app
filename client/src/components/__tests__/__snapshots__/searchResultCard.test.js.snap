// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchResultCard with all data > search result card exists with full data 1`] = `
<div class="pdap-grid-item flex flex-col border border-neutral-400 p-3 text-lg leading-snug">
  <h2 class="text-xl font-semibold line-clamp-2">Calls for Service for Cicero Police Department - IN</h2>
  <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Record type </p>
  <div class="mt-1 py-[.125rem] px-3 rounded-full bg-brand-wine/10 dark:bg-brand-wine dark:text-white w-fit">Calls for Service</div>
  <div class="search-result-agency">
    <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Agency </p>
    <p>Cicero Police Department - IN</p>
  </div>
  <!-- hiding place for now
		<div data-test="search-result-place">
			<p
				v-if="dataSource.municipality && dataSource.state_iso"
				data-test="search-result-place-state-municipality"
			>
				{{ dataSource.municipality }}, {{ dataSource.state_iso }}
			</p>
			<p v-else-if="dataSource.municipality" data-test="search-result-place-municipality">
				{{ dataSource.municipality }}
			</p>
			<p v-else-if="dataSource.state_iso" data-test="search-result-place-state">
				{{ dataSource.state_iso }}
			</p>
			<p v-else data-test="search-result-place-unknown">Location Unknown</p>
		</div>
-->
  <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Time range </p>
  <p>2016–2017</p>
  <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4">Formats available</p>
  <ul class="mb-4">
    <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">[</li>
    <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">'</li>
    <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">j</li>
    <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">s</li>
    <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">o</li>
    <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">n</li>
    <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">'</li>
    <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">,</li>
    <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit"> </li>
    <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">'</li>
    <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">p</li>
    <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">d</li>
    <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">f</li>
    <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">'</li>
    <li class="mt-1 py-[.125rem] px-3 rounded-full bg-slate-200 dark:bg-slate-600 w-fit">]</li>
  </ul>
  <div class="mt-auto flex flex-col">
    <button class="pdap-button pdap-button-primary text-lg font-medium px-4 py-1 mt-4 mb-2 lg:mx-0 max-w-full"> Visit data source <i class="fa fa-external-link"></i>
    </button>
    <button class="pdap-button pdap-button-secondary text-lg font-medium px-4 py-1 lg:mx-0 max-w-full"> View details </button>
  </div>
</div>
`;

exports[`SearchResultCard with coverage end but not start > search result card exists with coverage end but not start 1`] = `
<div class="pdap-grid-item flex flex-col border border-neutral-400 p-3 text-lg leading-snug">
  <h2 class="text-xl font-semibold line-clamp-2"></h2>
  <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Record type </p>
  <p> Unknown </p>
  <div class="search-result-agency">
    <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Agency </p>
    <p>Unknown</p>
  </div>
  <!-- hiding place for now
		<div data-test="search-result-place">
			<p
				v-if="dataSource.municipality && dataSource.state_iso"
				data-test="search-result-place-state-municipality"
			>
				{{ dataSource.municipality }}, {{ dataSource.state_iso }}
			</p>
			<p v-else-if="dataSource.municipality" data-test="search-result-place-municipality">
				{{ dataSource.municipality }}
			</p>
			<p v-else-if="dataSource.state_iso" data-test="search-result-place-state">
				{{ dataSource.state_iso }}
			</p>
			<p v-else data-test="search-result-place-unknown">Location Unknown</p>
		</div>
-->
  <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Time range </p>
  <p> Unknown start–2016</p>
  <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4">Formats available</p>
  <p> Unknown </p>
  <div class="mt-auto flex flex-col">
    <button class="pdap-button pdap-button-primary text-lg font-medium px-4 py-1 mt-4 mb-2 lg:mx-0 max-w-full" disabled> Visit data source <i class="fa fa-external-link"></i>
    </button>
    <button class="pdap-button pdap-button-secondary text-lg font-medium px-4 py-1 lg:mx-0 max-w-full"> View details </button>
  </div>
</div>
`;

exports[`SearchResultCard with coverage start but not end > search result card exists with coverage start but not end 1`] = `
<div class="pdap-grid-item flex flex-col border border-neutral-400 p-3 text-lg leading-snug">
  <h2 class="text-xl font-semibold line-clamp-2"></h2>
  <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Record type </p>
  <p> Unknown </p>
  <div class="search-result-agency">
    <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Agency </p>
    <p>Unknown</p>
  </div>
  <!-- hiding place for now
		<div data-test="search-result-place">
			<p
				v-if="dataSource.municipality && dataSource.state_iso"
				data-test="search-result-place-state-municipality"
			>
				{{ dataSource.municipality }}, {{ dataSource.state_iso }}
			</p>
			<p v-else-if="dataSource.municipality" data-test="search-result-place-municipality">
				{{ dataSource.municipality }}
			</p>
			<p v-else-if="dataSource.state_iso" data-test="search-result-place-state">
				{{ dataSource.state_iso }}
			</p>
			<p v-else data-test="search-result-place-unknown">Location Unknown</p>
		</div>
-->
  <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Time range </p>
  <p>2016–Unknown end </p>
  <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4">Formats available</p>
  <p> Unknown </p>
  <div class="mt-auto flex flex-col">
    <button class="pdap-button pdap-button-primary text-lg font-medium px-4 py-1 mt-4 mb-2 lg:mx-0 max-w-full" disabled> Visit data source <i class="fa fa-external-link"></i>
    </button>
    <button class="pdap-button pdap-button-secondary text-lg font-medium px-4 py-1 lg:mx-0 max-w-full"> View details </button>
  </div>
</div>
`;

exports[`SearchResultCard with missing data > search result card exists with missing data 1`] = `
<div class="pdap-grid-item flex flex-col border border-neutral-400 p-3 text-lg leading-snug">
  <h2 class="text-xl font-semibold line-clamp-2">Calls for Service for Cicero Police Department - IN</h2>
  <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Record type </p>
  <p> Unknown </p>
  <div class="search-result-agency">
    <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Agency </p>
    <p>Unknown</p>
  </div>
  <!-- hiding place for now
		<div data-test="search-result-place">
			<p
				v-if="dataSource.municipality && dataSource.state_iso"
				data-test="search-result-place-state-municipality"
			>
				{{ dataSource.municipality }}, {{ dataSource.state_iso }}
			</p>
			<p v-else-if="dataSource.municipality" data-test="search-result-place-municipality">
				{{ dataSource.municipality }}
			</p>
			<p v-else-if="dataSource.state_iso" data-test="search-result-place-state">
				{{ dataSource.state_iso }}
			</p>
			<p v-else data-test="search-result-place-unknown">Location Unknown</p>
		</div>
-->
  <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4"> Time range </p>
  <p> Unknown </p>
  <p class="text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4">Formats available</p>
  <p> Unknown </p>
  <div class="mt-auto flex flex-col">
    <button class="pdap-button pdap-button-primary text-lg font-medium px-4 py-1 mt-4 mb-2 lg:mx-0 max-w-full" disabled> Visit data source <i class="fa fa-external-link"></i>
    </button>
    <button class="pdap-button pdap-button-secondary text-lg font-medium px-4 py-1 lg:mx-0 max-w-full"> View details </button>
  </div>
</div>
`;
